### ISSUE TITLE
develop new GUI using Electron

### ISSUE THREAD

**Issue by davidpanderson** [BOINC/boinc#3185]
Electron (https://electronjs.org/) is a possible framework for developing a new GUI.
It's cross-platform, uses HTML for layout and CSS for styling.
It seems to be widely used and well-supported.

If anyone is interested in working on this, please let me know.

BTW, the BOINC client exposes GUI RPCs via HTTP;
I've verified that these work with the HTTP and XML features built into Javascript.

**Comment by lfield** [comment]
This is similar to issues #2112, #2113 and PR #2275. Even though the BOINC client exposes GUI RPCs via HTTP, it is not possible to use this directly from a browser. As far as I understand Electron uses Chromium so will have the same issue. To make this work the client needs to provide a Web socket. There are some security concerns related to this that have been discussed in PR #2275. I believe that it is possible to do this but have not had time to make any progress but could do so if we consider this a direction we would like to go.

**Comment by adamradocz** [comment]
I'm working on a complete rewrite of the [Boinc Manager](https://github.com/adamradocz/boinc-manager). It uses a shared code (.NET Standard) and some platform-specific code. It will work on:
- iOS, Android (Xamarin)
- Windows (.NET Core WPF)
- Web (ASP.NET Core).

[Electron.NET](https://github.com/ElectronNET/Electron.NET) could be a good candidate for Linux and Mac.

**Comment by davidpanderson** [comment]
Electron doesn't have the RPC security issue.

**Comment by lfield** [comment]
Electron might not but to make it communicate with the BOINC client you need to add a WebSocket. I remember us discussing this during the hackathon at the workshop in Paris. As soon as you do this you open the client to JS drive by attacks. Hence we have to be quite sure regarding the security considerations of this approach. It is one of the reasons why PR  #2275 was closed rather than merged.

**Comment by davidpanderson** [comment]
I have an Electron app that communicates with the BOINC client over TCP, just like the WxWidgets GUI does.  Take my word for it.

**Comment by lfield** [comment]
That's interesting. I wonder how it works. Do you have any sample code to try?

**Comment by UweBeckert** [comment]
Electron creates a stand alone app. The chromium inside electron is a browser just for rendering html/css and running javascript. It doesn't has the "same origin policy" and other security feature which a normal browser has. I wrote already electron applications which communicate via http to different server or which run soap operations and also parses xml.

**Comment by CyberTailor** [comment]
If BOINC Manager uses Electron, it'll become too bloated and, moreover, it'll be harder to make packages for Linux distros (\*sad gentoo user sounds\*)

**Comment by davidpanderson** [comment]
The current GUI, based on WxWidgets, is about as bloated as you can get, in terms of source code which is what matters.  Electron could reduce this by a factor of 10.

**Comment by ScyllaHide** [comment]
> 
> 
> If BOINC Manager uses Electron, it'll become too bloated and, moreover, it'll be harder to make packages for Linux distros (*sad gentoo user sounds*)

yes thats my thought too! but on the other hand i know not to much about the current old gui BM has.

**Comment by StaticallyTypedRice** [comment]
Electron is a bad idea. It's bloated as others have mentioned, and it's also extremely resource hungry. Worse, it relies on Chromium which Google controls, and if I recall correctly has been shown to have privacy issues (like phoning home to Google) even though it's not full Google Chrome. I implore you to use something else, like GTK, Qt, one of the Java/Kotlin GUI libraries. Electron is a terrible GUI library.

**Comment by StaticallyTypedRice** [comment]
> The current GUI, based on WxWidgets, is about as bloated as you can get, in terms of source code which is what matters. Electron could reduce this by a factor of 10.

At the cost of significantly higher resource usage on the user's computer and possible privacy issues with the embedded Chromium instance.

**Comment by fturco** [comment]
Please don't use Electron. GTK or Qt are both way better than Electron.

**Comment by AenBleidd** [comment]
@fturco, what are your arguments?

**Comment by fturco** [comment]
Well... I already have GTK and Qt installed on my Gentoo Linux system, and I prefer not installing other packages, especially if they require long build times.

**Comment by CyberTailor** [comment]
@fturco, GTK isn't truly cross-platform so it isn't an option. Also there are more important tasks than reimplementing GUI.

**Comment by RokeJulianLockhart** [comment]
> https://github.com/BOINC/boinc/issues/3185#issuecomment-636077288

Yeah, @CyberTailor, GTK4 wouldn't be a good choice. It's as close to CSD-only as you can get with a GUI toolkit, and doesn't support theming whatsoever (which is dreadful for accessibility).

> https://github.com/BOINC/boinc/issues/3185#issuecomment-614295276

@davidpanderson, is that *due* to WxWidgets? I imagine not, so would it not be easier to merely remake the WxWidgets app?

To my knowledge, if you're trying to make a cross-platform, performant GUI, you want Qt6 QtWidgets (maybe QML is quick too, but IDK) or WxWidgets (my LINX 1010's EFI GUI is written in it).

**Comment by AenBleidd** [comment]
@RokeJulianLockhart, with Electron you can make much nicer GUI in a significantly easier way. Also, this will give us a possibility to introduce WebGUI that was also requested by users.
So no, we don't want to rewrite GUI neither with wxWidgets nor with Qt.

**Comment by RokeJulianLockhart** [comment]
> https://github.com/BOINC/boinc/issues/3185#issuecomment-2531450500

@AenBleidd, I'm fairly impartial to each option, but I'm surprised that everyone here offers Electron as the sole method by which they can implement a web technology-based GUI. Does it offer something that, for example, CEF (think Steam) or `QtWebEngine` (RStudio) do not?

**Comment by CyberTailor** [comment]
> > [#3185 (comment)](https://github.com/BOINC/boinc/issues/3185#issuecomment-2531450500)
> 
> [@AenBleidd](https://github.com/AenBleidd), good to hear!
> 
> I'm fairly impartial to each option, but I'm surprised that everyone here offers Electron as the sole method by which they can implement a web technology-based GUI. Does it offer something that, for example, CEF (think Steam) or `QtWebEngine` (RStudio) do not?

[Tauri](https://tauri.app/) is "Electron but better". It uses OS's native web renderer instead of making each application carry its own copy of Chromium.

Though either Tauri or Electron would make it basically impossible for many Linux distributions to make own packages of BOINC manager, and they'll have no other option than using upstream-provided prebuilt packages.

**Comment by AenBleidd** [comment]
> Though either Tauri or Electron would make it basically impossible for many Linux distributions to make own packages of BOINC manager, and they'll have no other option than using upstream-provided prebuilt packages.

@CyberTailor, why?

**Comment by AenBleidd** [comment]
> Does it offer something that, for example, CEF (think Steam) or QtWebEngine (RStudio) do not?

You can make a research and provide us with a comparison why any of your options better than ours.
At the moment we created this ticket and basically came to this idea, Electron was (and still is I think) modern and popular way of creating nice looking applications (e.g. VSCode and many others). So for us this is still the best option.

**Comment by RokeJulianLockhart** [comment]
> You can make a research and provide us with a comparison why any of your options better than ours.

@AenBleidd, I *didn't* think they were, until I saw @CyberTailor's comment: [^1]

> [Tauri](https://tauri.app/) is "Electron but better". It uses OS's native web renderer instead of making each application carry its own copy of Chromium.

That's a significant advantage. The embedded chromium in Electron applications can quite easily become out-of-date if the packager doesn't keep atop their update regimen, because unlike on AOSP and i(Pad)OS, there is no WebView component to keep it updated.

Additionally, https://github.com/electron/electron/issues/41551#issue-2175565420 remains open, which means that Electron applications must be invoked via XWayland on Linux-based OSes, which introduces another layer of instability (and slowness, very slightly).

[^1]: https://github.com/BOINC/boinc/issues/3185#issuecomment-2531555132

**Comment by RokeJulianLockhart** [comment]
> https://github.com/BOINC/boinc/issues/3185#issuecomment-2531574752

@CyberTailor, it wouldn't make packaging as difficult as you state. You've mentioned at https://github.com/BOINC/boinc/issues/3185#issuecomment-614254696 that you utilize Gentoo, so you should expect to spend your life compiling dependencies!

Multiple Electron-based applications are packaged for Fedora 41. It's of no consequence, @AenBleidd.

**Comment by AenBleidd** [comment]
> Additionally, https://github.com/electron/electron/issues/41551#issue-2175565420 remains open, which means that Electron applications must be invoked via XWayland on Linux-based OSes, which introduces another layer of instability (and slowness, very slightly).

This is not something we are working on at the moment.
So there is a big possibility that when we will actually start developing it, this might be already solved or we can even consider using completely different framework.
As you can see, this ticket was open more than 5 years ago, and we haven't started working on it yet.

Also, this is a very low priority ticket, since we have currently a GUI that is stable and functioning.
Yes, we want to get rid of it in favor of something more popular and easy to develop, but such kind of work will require a lot of resources and this is something we don't have at the moment.

So this also mean that right now this discussion will not move this ticket in any direction.

**Comment by Germano0** [comment]
> Multiple Electron-based applications are packaged for Fedora 41

as Fedora package maintainer and after having had a talk with some other fellow Fedora devs, I can tell you that Fedora is NOT shipping any Electron based application on its repository because nobody has done the work to make a packaged electron runtime.

Hereunder I paste some messages (with the permission of the author) extracted from a talk with from Neal Gompa (@conan-kudo), one of the most important Fedora developers, who kindly wrote the following statements regarding this boinc-client GUI discussion

```
-------
```
if cross-platform is what they want, they should consider Qt even the Android GUI could be written in Qt if they use the Kirigami stuff.
There's a Matrix room aimed to help people learn about developing applications using KDE technologies if you want to forward them to it: 
https://matrix.to/#/#kde-streamlined-app-dev:kde.org
Here's a website about KDE Frameworks too: https://develop.kde.org/products/frameworks/

Qt + KDE Frameworks gives you a leg up on a lot of common patterns and interfaces. It simplifies development and gives you clean, ABI stable extensions on top of Qt to support application development.

Usage of KDE frameworks does not imply that it's a KDE application, and indeed you can have whatever UX or platform preferences you want even with it.

Things like config files (KConfig), push notifications (KUnifiedPush), status icons/appindicators (KStatusNotifierItem), clean window layouts and extra widgets (Kirigami and KWidgetsAddons), network detection (NetworkManagerQt), idle detection (KIdleTime), gettext-style translations (K18n), and others, it becomes much easier to implement a consistent and useful user experience across Windows, Mac, Linux/Unix, and Android.

The frameworks and platform support for each framework module is listed here: https://api.kde.org/frameworks/index.html
The benefit here over wxWidgets is that Qt is a complete platform abstraction layer, so the GUI codebase can be much simpler than it is now and there's a lot of people developing with Qt all over, with professional documentation, tutorials, and examples to work from

Qt has the ability to interface with web-based protocols too, so client-server interactions should be pretty straightforward.
I've seen applications use RESTful HTTP, XML-RPC, JSON-RPC, and gRPC fairly easily with the stuff provided by Qt itself

**Comment by RokeJulianLockhart** [comment]
> I can tell you that Fedora is NOT shipping any Electron based application on its repository because nobody has done the work to make a packaged electron runtime.

@Germano0, I was thinking of RPMFusion. Apologies.

> If cross-platform is what they want, they should consider Qt even the Android GUI could be written in Qt if they use the Kirigami stuff.

I'll affirm this. However, keep https://invent.kde.org/frameworks/kirigami/-/issues/81 in mind.

**Comment by RokeJulianLockhart** [comment]
> https://github.com/BOINC/boinc/issues/3185#issuecomment-2536365697

@AenBleidd, I understand you being dead-set on using a web technology GUI framework, considering the significant reduction in effort to make a cross-platform application that way.

However, why are you also apparently dead-set on Electron? Another user previously mentioned Tauri, which appeared to be an improvement (in all respects that I've researched).

> Anyway, we don't need to rely on any OS, and can include Electron engine inside the application installation package (like this is done by vscode for example).

Yeah, I think that's what all applications currently packaged as RPMs do, whether built on RPMFusion (like Discord) or not (like VS Code, Stable and Insiders).

**Comment by AenBleidd** [comment]
> However, why are you also apparently dead-set on Electron?

I'm not dead-set. I'm open for suggestions.
I just see no reason to discuss it now since we don't plan doing any development in the near future.

**Comment by RokeJulianLockhart** [comment]
> I just see no reason to discuss it now since we don't plan doing any development in the near future.

@AenBleidd, unless we were discussing the limitations of very immature or soon-to-be-deprecated technologies, I'd agree. However, since we're discussing fairly consistent staples of the current development landscape, this time might be advantageous for us to properly settle on something.

I do understand, though, and talking about talking probably is a waste of time, so I'll shut up about *this.*

**Comment by homersimpsons** [comment]
The advantage of using a web-based ui framework is that it is easy to translate it to a browser usage. But such solutions are often costly in terms of performances, notably in RAM.

Maybe we could use a lower level technology such as https://slint.dev/ that [supports C++](https://docs.slint.dev/latest/docs/cpp/).

(Edit, to avoid spamming the thread) For reference:
- LibrePCB switched incrementally from Qt to Slint: https://librepcb.org/blog/2025-09-12_preview_of_next_gen_ui/
- Slint release material template: https://material.slint.dev/

**Comment by davidpanderson** [comment]
I'd be interested in seeing a minimal example - maybe just lists of projects and tasks -
using Slint or similar technology.

But as Vitalii says, this would be a 3rd party effort, like BoincTasks;
we don't have the resources to participate in or manage it.

**Comment by homersimpsons** [comment]
> I'd be interested in seeing a minimal example - maybe just lists of projects and tasks - using Slint or similar technology.

Since I wrote my message (1h ago) I tried to play with it a bit (I have no prior experience with it on https://www.slintpad.com/) and could replicate the project list (I believe the rendering issues is only on the web):

![Image](https://github.com/user-attachments/assets/79095811-3ffe-4db8-a813-d8a807129117)

Of course nothing is wired (I have little to no experience in C++) but the code is currently pretty concise:

<details>
  <summary>Slint code</summary>
  
```slint
import { AboutSlint, Button, VerticalBox, TabWidget, StandardTableView, HorizontalBox } from "std-widgets.slint";

export component Demo inherits Window {
    title: "BOINC Manager";
    no-frame: false;
    preferred_height: 550px;
    preferred_width: 920px;
    resize-border-width: 1px;

    VerticalLayout {
        HorizontalBox {
            alignment: LayoutAlignment.start;
            horizontal-stretch: 0;
            spacing: 10px;
            Text { text: @tr("File"); }
            Text { text: @tr("View"); }
            Text { text: @tr("Activity"); }
            Text { text: @tr("Options"); }
            Text { text: @tr("Tools"); }
            Text { text: @tr("Help"); }
        }

        TabWidget {
            Tab {
                title: @tr("Notifications");
            }

            Tab {
                title: @tr("Projects");
                HorizontalBox {
                    VerticalBox {
                        alignment: LayoutAlignment.start;
                        width: 150px;
                        Button { text: @tr("Update"); }
                        Button { text: @tr("Suspend"); }
                        Button { text: @tr("Allow new tasks"); }
                        Button { text: @tr("Reset project"); }
                        Button { text: @tr("Remove Project"); }
                        Button { text: @tr("Properties"); }
                     }
                    StandardTableView {
                        columns: [
                            { title: @tr("Project") },
                            { title: @tr("Account") },
                            { title: @tr("Team") },
                            { title: @tr("Work done") },
                            { title: @tr("Avg. work done") },
                            { title: @tr("Resource share") },
                            { title: @tr("Status") },
                        ];
                        // Project List
                        rows: [
                            [
                                { text: "World Community Grid" }, { text: "homer_simpsons" }, { text: "L'Alliance Francophone" }, { text: "3 784 233" }, { text: "1 121,21" }, { text: "500" }, { text: "" },
                            ],
                        ];
                     }
                }
            }

            Tab {
                title: @tr("Tasks");
            }

            Tab {
                title: @tr("Transfers");
            }

            Tab {
                title: @tr("Statistics");
            }

            Tab {
                title: @tr("Storage");
            }
        }
        HorizontalBox {
            padding: 5px;
            Text { text: @tr("Connected to {} ({})", "localhost", "8.0.2");}
        }
    }
}
```
  
</details>

**Comment by AenBleidd** [comment]
We need to use commonly-used technologies. There are new frameworks that appear every day.
We don't have a luxury to jump to a new framework every time it appears.
That is why the original idea was regarding Electron that is widely used.

**Comment by davidpanderson** [comment]
How would one do the GUI RPCs to get the project / task info from the client?

**Comment by AenBleidd** [comment]
> How would one do the GUI RPCs to get the project / task info from the client?

That's the most important: GUI is not only the UI but also some logic behind.
The example above looks simple, true, but it's just a mockup without any real workflow (even the simple one).

I suggest not to spend time on that unless there is a group of people who wants to develop and maintain this project in a long term.
But even in that case we need to do the investigation one more time to evaluate available technologies and choose the one that fits our goals the best. It might be even not the Electron at the end but something else.

**Comment by AenBleidd** [comment]
A nice 3rd-party GUI was developed.
Default BOINC installation should include small, lightweight implementation with the basic set of functionality.
Since BOINC Manager might run on a low end devices with limited resources, there is no good reason to replace the current GUI with something based on Electron on amy other similar technology.
And we have no resources to support several GUIs at the same time.
We might revisit this in future, but as for now I'm closing this as 'not planned'.
